### PR TITLE
Change `styleCheck` to ignore foreign packages

### LINKED ISSUE #10201: [styleCheck] --styleCheck:error inconsistent with --styleCheck:hint : doesn't respect package boundaries
--styleCheck:hint inconsistent with --styleCheck:error:
i would expect `--styleCheck:hint` to report exact same issues as --styleCheck:error but treating them as hints instead of errors, however:

```nim
proc main_bar() = discard
```

```
nim c --styleCheck:hint $timn_D/bugs/all/t0104.nim
Hint: used config file '/Users/timothee/git_clone/nim/Nim/config/nim.cfg' [Conf]
Hint: used config file '/Users/timothee/.config/nim/nim.cfg' [Conf]
Hint: used config file '/Users/timothee/git_clone/nim/timn/nim.cfg' [Conf]
Hint: used config file '/Users/timothee/git_clone/nim/Nim/config/config.nims' [Conf]
/Users/timothee/git_clone/nim/timn/bugs/all/t0104.nim(9, 6) Hint: name should be: 'mainBar' [Name]
  proc main_bar() = discard
       ^
Hint: operation successful (29804 lines compiled; 0.363 sec total; 25.137MiB peakmem; Debug Build) [SuccessX]
```

```
nim c --styleCheck:error $timn_D/bugs/all/t0104.nim
Hint: used config file '/Users/timothee/git_clone/nim/Nim/config/nim.cfg' [Conf]
Hint: used config file '/Users/timothee/.config/nim/nim.cfg' [Conf]
Hint: used config file '/Users/timothee/git_clone/nim/timn/nim.cfg' [Conf]
Hint: used config file '/Users/timothee/git_clone/nim/Nim/config/config.nims' [Conf]
../../../Nim/lib/system/ansi_c.nim(110, 6) Error: name should be: 'CSighandlerT'
```

### PR DESCRIPTION
**Pull request by quantimnot**
* Symbols from foreign packages are now ignored.
* Fixed `styleCheck` violations in `compiler` package.
* Added symbol ownership to custom annotation pragmas.
* Minor refactors to cleanup style check callsites.
* Minor internal documentation of reasons why a symbol isn't checked.

Style violations were fixed in the compiler after thet were exposed by
the changes. The compiler wouldn't compile otherwise.

Symbol ownership for custom pragma annotations is needed for checking
the annotation's style. A NPE was raised otherwise.

Fixes #10201
See also nim-lang/RFCs#456

### CODE REVIEW

**Comment by quantimnot** [comment]
This is ready for review.

File: compiler/linter.nim [hunk 1]
```diff
@@ -116,23 +125,31 @@ proc differs(conf: ConfigRef; info: TLineInfo; newName: string): string =
...
+proc styleCheckUseImpl(conf: ConfigRef; info: TLineInfo; s: PSym) =
   let newName = s.name.s
   let badName = differs(conf, info, newName)
   if badName.len > 0:
     # special rules for historical reasons
-    let forceHint = badName == "nnkArgList" and newName == "nnkArglist" or badName == "nnkArglist" and newName == "nnkArgList"
-    lintReport(conf, info, newName, badName, forceHint = forceHint, extraMsg = "".dup(addDeclaredLoc(conf, s)))
-
-proc checkPragmaUse*(conf: ConfigRef; info: TLineInfo; w: TSpecialWord; pragmaName: string) =
+    let forceHint = # TODO: PR Reviewer: Any clue if this is still needed? (This comment needs removed before commit.)
```
Comment (Araq) [inline compiler/linter.nim]: Probably still required for bootstrapping but it really sucks.

Comment (quantimnot) [reply compiler/linter.nim]: Ok, thank you. I found the PR (#17529) that introduced it. I removed it and was still able to bootstrap the compiler. So, considering that it boots and that this PR prevents the style check in foreign packages, it seems useless and is now removed.

**Comment by Araq** [comment]
Ping @arnetheduck. Does this need to be backported?

**Comment by arnetheduck** [comment]
> Ping @arnetheduck. Does this need to be backported?

yes pls

**Comment by omentic** [comment]
Sorry to necro this, but does "Symbols from foreign packages are now ignored" mean that inconsistent usages of imported symbols sneaks by? ex.

```nim
import std/os
discard fileExists("hello")
discard file_exists("hello") # inconsistency not noticed
```

**Comment by quantimnot** [comment]
@j-james
You are correct. I hadn't considered that situation. I will make a new PR, or tack it onto #20095 since it hasn't been reviewed yet.

**Pull request merged**